pylint/checkers/method_args.py

```diff
@@ -0,0 +1,88 @@
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt
+
+"""Variables checkers for Python code."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from astroid import nodes
+
+from pylint.checkers import BaseChecker, utils
+from pylint.interfaces import INFERENCE
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+
+class MethodArgsChecker(BaseChecker):
+    """BaseChecker for method_args.
+
+    Checks for
+    * missing-timeout
+    """
+
+    name = "method_args"
+    msgs = {
+        "W3101": (
+            "Missing timeout argument for method '%s' can cause your program to hang indefinitely",
+            "missing-timeout",
+            "Used when a method needs a 'timeout' parameter "
+            "in order to avoid waiting for a long time. If no timeout "
+            "is specified explicitly the default value is used, for example "
+            "for 'requests' the program will never time out (i.e. hang indefinitely). "
+            "Notice the libraries needs to be installed and the name should be the output of pylint infered. "
+            "from astroid import extract_node;"
+            "node = extract_node('from requests import get;get()');"
+            "print(next(node.func.infer()).qname())",
+        ),
+    }
+    options = (
+        (
+            "timeout-methods",
+            {
+                "default": (
+                    "requests.api.delete",
+                    "requests.api.get",
+                    "requests.api.head",
+                    "requests.api.options",
+                    "requests.api.patch",
+                    "requests.api.post",
+                    "requests.api.put",
+                    "requests.api.request",
+                ),
+                "type": "csv",
+                "metavar": "<comma separated list>",
+                "help": "List of qualified names (i.e., library.method) which require a timeout parameter "
+                "e.g. 'requests.api.get,requests.api.post'",
+            },
+        ),
+    )
+
+    @utils.only_required_for_messages(
+        "missing-timeout",
+    )
+    def visit_call(self, node: nodes.Call) -> None:
+        """Check if the call needs a timeout parameter based on package.func_name
+        configured in config.timeout_methods.
+
+        Package uses inferred node in order to know the package imported.
+        """
+        inferred = utils.safe_infer(node.func)
+        if inferred and inferred.qname() in self.config.timeout_methods:
+            for keyword in node.keywords:
+                if isinstance(keyword, nodes.Keyword) or keyword.arg == "timeout":
+                    break
+            else:
+                self.add_message(
+                    "missing-timeout",
+                    node=node,
+                    args=(node.func.as_string(),),
+                    confidence=INFERENCE,
+                )
```

Comment:
Say we want to catch call of ``whatever.bar`` and ``sentry.something.foo`` that do not define a ``foobar`` argument, and then raise a ``missing-foobar`` message, we'd have to refactor the checker for it to work. The checker's current name imply this could be a goal. We could also rename the checker to ``MissingTimeoutChecker`` and not handle the generic case, I wouldn't mind, but I think it was nice to identify that it could be generic.

Reply:
It is expected

A new checker needs to change a lot of things

One of them is create the new checker number, doc, unittest and change the logic of the method to support it

It can not be configured dynamically since that is not possible to add new checkers dynamically as far I know

Reply:
We could add values to check the argument x is defined in functions, f1, f2, f3, by modifying a data structure instead of adding new code. Say something like this:

```diff
        [
            {
                "functions": (
                    "requests.api.delete",
                    "requests.api.get",
                    "requests.api.head",
                    "requests.api.options",
                    "requests.api.patch",
                    "requests.api.post",
                    "requests.api.put",
                    "requests.api.request",
                ),
                "required_args": "timeout",
            },
+           {
+               "functions": (
+                   "whatever.bar",
+                   "sentry.something.foo",
+               ),
+               "required_args": "foobar",
+           }
        ]
```

But let's just rename the checker.

Reply:
I got it

Thank you for clarify

I even don't got you

How will you match that `timeout` is for requests and foobar is for `whatever.bar`?

Could you share me a checker using this way, please?

Reply:
Another question

> We could also rename the checker to MissingTimeoutChecker 

Do you mean, Using only one checker by file?

Is this a good idea?

I mean, if you have a guideline to define a checker by file you could have too many files in the future instead of re-use the same checker to add new ones in the future

Similar to imports.py, variables.py, strings.py... and so on

Using your guideline it should be:
 - bad_format_character.py

because it was the first check

Then add `truncated_format_string.py`

and so on

IMHO It is not a good idea if you can re-use a file to add many checkers

Even using this way the limit of 99 checkers prefix could be reached faster

Is not?

Reply:
> How will you match that timeout is for requests and foobar is for whatever.bar?

They are in a different structure, we can iterate over the list of structure and get ``functions`` and  ``required_args``for each.  Please note that it's only an illustration the option and message name and description are not included there's some design to do before it works.

> Could you share me a checker using this way, please?

This would be a novel approach I guess. But the deprecated checker is pretty modular for example, you can define deprecated functions, and programmatically check that a particular function is deprecated by inheriting the checker (to check for example the version of django and have different result based on this).

>  It is not a good idea if you can re-use a file to add many checkers

There's a balance to have between mixing many messages in the same file and having strange bug when they interact strangely with each other, or having a lot of independent files. I'm not too afraid of having too many files and having to add a zero in the message id, but grouping is also nice if it's done in a way that is logical and easy to follow.

It could go both way, if you want to do a generic checker we can do that, but if you just want to only do a missing timeout checker it works too.

Reply:
Maybe using

library.method:parameter?

Reply:
Hmm yeah that would work, but I don't think we want all the function/parameter to be mixed and the message to be a very generic "missing-argument". We want to explain why it's better to have timeout defined (hanging indefinitely otherwise) and we don't know yet why ``foobar`` in ``libary.method:parameter`` need to be defined.

Reply:
Sorry I was away yesterday so I haven't completely followed the discussion. I don't think this option should be generic: that will become unnecessarily confusing for users.

I do think that the checker class can be generic. If we find new arguments we want to check for we can just add a new option.

Reply:
Totally agree

So, Let me know if it requires a change, please

Reply:
> I do think that the checker class can be generic. 

Let's do this in a follow-up.

Reply:
My bad, I meant `checker` function. Let's not start adding abstract generic checker classes 😅

Reply:
Sorry, I even don't get what should I change here

Could you help me to understand the changes required, please?

Reply:
The idea is to have a generic function so we could add data to a structure to check similar cases by modifying the data structure and not adding code (https://github.com/PyCQA/pylint/pull/6780#discussion_r894875851). But let's design it and refactor in another PR, this one is already doing a lot.